### PR TITLE
Editor 3.0

### PR DESCRIPTION
**Pull request by juliankrispel**
Changes:

- [x] flow types everywhere (almost done)
- [x] updated tests
- [x] refactored hook resolution (no more magic matching patterns)
- [x] export flow types in package
- [x] decorator resolution (deprecate editorState.decorator)
- [x] add build step for flow types

## Checklist

- [x] Fix any eslint errors that occur
- [x] Update change-log for every plugin you touch
- [x] Add/Update tests if you add/change new functionality
- [ ] Add/Update docs if you add/change functionality

### CODE REVIEW

**Comment by mxstbr** [comment]
Let me know when an alpha is out and I'll test it!

**Comment by juliankrispel** [comment]
@mxstbr I've added a story `DynamicPluginEditor` if you can test it in there and try break the thing that'd be awesome!

File: draft-js-plugins-editor/src/Editor/__test__/index.js [hunk 1]
```diff
@@ -671,20 +684,5 @@ describe('Editor', () => {
       expect(customPluginDecorator).has.been.called();
       expect(decoratorStrategy).has.been.called();
     });
-
-    it('reassigns decorators to editorState when props are updated with naked editorState', (done) => {
```
Comment (juliankrispel) [inline draft-js-plugins-editor/src/Editor/__test__/index.js]: this is now deprecated

File: draft-js-plugins-editor/src/Editor/MultiDecorator.js [hunk 2]
```diff
@@ -46,11 +46,8 @@ class MultiDecorator {
 
   /**
    * Return props to render a decoration
-   *
-   * @param {String} key
-   * @return {Object}
    */
-  getPropsForKey(key) {
+  getPropsForKey(key: string): ?Object {
```
[the remark this answers is not in the document]
Comment (juliankrispel) [reply draft-js-plugins-editor/src/Editor/MultiDecorator.js]: I'm actually copying draft-js internal types here - to be consistent

File: draft-js-plugins-editor/package.json [hunk 3]
```diff
@@ -23,15 +23,19 @@
...
     "prepublish": "npm run build"
   },
   "license": "MIT",
   "dependencies": {
     "decorate-component-with-props": "^1.0.2",
     "find-with-regex": "^1.1.3",
+    "flow-copy-source": "^1.3.0",
     "immutable": "~3.7.4",
     "prop-types": "^15.5.8",
+    "shallowequal": "^1.0.2",
```
Comment (nikgraf) [inline draft-js-plugins-editor/package.json]: I'm wondering if this dependency is necessary. I'm not sure if React or Draft has a dependency on fbjs (https://www.npmjs.com/package/fbjs). If that's the case we can use the shallowequal from  that without introducing duplicated code.

File: draft-js-plugins-editor/src/Editor/createCompositeDecorator.js [hunk 4]
```diff
@@ -1,18 +1,23 @@
...
+  decorators: List<DraftDecorator>,
+  getEditorState: () => EditorState,
+  setEditorState: (editorState: EditorState) => void) => {
+  const convertedDecorators = decorators
     .map((decorator) => ({
       ...decorator,
       component: decorateComponentWithProps(decorator.component, { getEditorState, setEditorState }),
-    }))
-    .toJS();
+    })).toArray();
```
Comment (nikgraf) [inline draft-js-plugins-editor/src/Editor/createCompositeDecorator.js]: I'm not 100% sure, but afaik toJS does a deep conversion and toArray only a shallow one. Is this desired here?

File: draft-js-plugins-editor/src/Editor/index.js [hunk 5]
```diff
@@ -74,62 +92,88 @@ class PluginEditor extends Component {
...
   // Cycle through the plugins, changing the editor state with what the plugins
   // changed (or didn't)
-  onChange = (editorState) => {
+  onChange = (editorState: EditorState) => {
     let newEditorState = editorState;
+    const pluginMethods = this.getPluginMethods();
     this.resolvePlugins().forEach((plugin) => {
-      if (plugin.onChange) {
-        newEditorState = plugin.onChange(newEditorState, this.getPluginMethods());
+      if (plugin.onChange != null && newEditorState != null) {
```
Comment (nikgraf) [inline draft-js-plugins-editor/src/Editor/index.js]: so if I call onChange with `null` it will not trigger the onChange callbacks for plugins. Is this desired behaviour?

File: draft-js-plugins-editor/src/Editor/index.js [hunk 6]
```diff
@@ -74,62 +92,88 @@ class PluginEditor extends Component {
...
+        newEditorState = plugin.onChange(newEditorState, pluginMethods);
       }
     });
 
     if (this.props.onChange) {
```
Comment (nikgraf) [inline draft-js-plugins-editor/src/Editor/index.js]: Especially compared to above we have different behavior to plugins since this onChange would still fire even though the new editor state is `null`

File: draft-js-plugins-editor/src/Editor/index.js [hunk 7]
```diff
@@ -139,165 +183,197 @@ class PluginEditor extends Component {
...
+  createHandleHooks = (methodName: string, plugins: Array<Plugin|EditorProps>) => (firstArg: any, ...args:any):string => {
+    const newArgs = [firstArg, ...args, this.getPluginMethods()];
 
-    return plugins.some((plugin) =>
-      typeof plugin[methodName] === 'function'
-        && plugin[methodName](...newArgs) === 'handled'
-    ) ? 'handled' : 'not-handled';
+    return plugins.some((plugin) => {
+      const method = plugin[methodName];
+      return method != null && method(...newArgs) === 'handled';
```
Comment (nikgraf) [inline draft-js-plugins-editor/src/Editor/index.js]: why not use typeof here? Not null could still mean it's not a function.

File: stories/DynamicPluginsEditor/index.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
+import React, { Component } from 'react';
```
Comment (nikgraf) [inline stories/DynamicPluginsEditor/index.js]: cool feature the DynamicPlugins!

File: package.json [hunk 9]
```diff
@@ -103,7 +104,7 @@
     "emojione": "^2.2.7",
     "eslint-plugin-import": "^2.2.0",
     "find-with-regex": "^1.1.3",
-    "immutable": "^3.8.1",
+    "immutable": "4.0.0-rc.9",
```
Comment (nikgraf) [inline package.json]: does it make sense to wait until immutable actually released the 4.0?

File: draft-js-plugins-editor/src/Editor/resolveDecorators.js [hunk 10]
```diff
@@ -1,27 +1,47 @@
...
 
 // Return true if decorator implements the DraftDecoratorType interface
 // @see https://github.com/facebook/draft-js/blob/master/src/model/decorators/DraftDecoratorType.js
-const decoratorIsCustom = (decorator) => typeof decorator.getDecorations === 'function' &&
-  typeof decorator.getComponentForKey === 'function' &&
-  typeof decorator.getPropsForKey === 'function';
-
+// $FlowFixMe
+const decoratorIsCustom = (decorator: DraftDecorator|DraftDecoratorType): boolean => (
+  decorator.getDecorations != null &&
```
Comment (nikgraf) [inline draft-js-plugins-editor/src/Editor/resolveDecorators.js]: is this flow related? checking if it's a function was on purpose back then :)

File: draft-js-plugins-editor/src/Editor/index.js [hunk 11]
```diff
@@ -139,165 +183,197 @@ class PluginEditor extends Component {
...
-    const plugins = [this.props, ...this.resolvePlugins()];
+    const plugins = this.resolvePlugins();
+
     plugins.forEach((plugin) => {
       if (typeof plugin.getAccessibilityProps !== 'function') return;
       const props = plugin.getAccessibilityProps();
       const popupProps = {};
 
-      if (accessibilityProps.ariaHasPopup === undefined) {
+      if (accessibilityProps.ariaHasPopup == null) {
```
Comment (nikgraf) [inline draft-js-plugins-editor/src/Editor/index.js]: Is this every true? I guess most of the times it's undefined and that's why we had the check for it?!

File: draft-js-plugins-editor/package.json (hunk 3 above)
Comment (mxstbr) [reply draft-js-plugins-editor/package.json]: I'm pretty sure React depends on fbjs, or at least it did a while ago

File: draft-js-plugins-editor/src/Editor/index.js (hunk 11 above)
Comment (mxstbr) [reply draft-js-plugins-editor/src/Editor/index.js]: `value == null` is a shorthand for `value === null || value === undefined` :+1: #javascript

File: draft-js-plugins-editor/package.json (hunk 3 above)
Comment (juliankrispel) [reply draft-js-plugins-editor/package.json]: it's not actually, I had used it and then forgot to take it out ....

**Comment by juliankrispel** [comment]
@nikgraf thx for the review - yes a bunch of the changes were to satisfy flow checks to be fair. I'll see if I can fix that to some degree. It should really give us more security rather than less...

File: draft-js-plugins-editor/src/Editor/index.js (hunk 11 above)
Comment (nikgraf) [reply draft-js-plugins-editor/src/Editor/index.js]: ah missed the `==` … can you add that as a comment on top of it to make it beginner friendly and people like me constantly forgetting this :)

**Comment by nikgraf** [comment]
I'm approving in the meantime. Ping me if you want to have another review after more changes :)

**Comment by sibelius** [comment]
This is great

I'm gonna close this until we refactor most of the code to bring new changes
